networking/src/quic/endpoint/mod.rs

```diff
@@ -0,0 +1,307 @@
+//! Starting point to create a QUIC enabled network socket.
+
+mod transport;
+
+use std::{
+    fmt::{self, Debug, Formatter},
+    net::{SocketAddr, ToSocketAddrs},
+    pin::Pin,
+    sync::Arc,
+    task::{Context, Poll},
+};
+
+use flume::{r#async::RecvStream, Sender};
+use futures_channel::oneshot;
+use futures_util::{stream::Stream, StreamExt};
+use quinn::{
+    CertificateChain, ClientConfig, ClientConfigBuilder, NewConnection, ServerConfig,
+    ServerConfigBuilder, VarInt,
+};
+
+use super::{StreamExtExt, Task};
+use crate::{
+    certificate::{Certificate, PrivateKey},
+    Connection, Error, Result,
+};
+
+/// Represents a socket using the QUIC protocol to communicate with peers.
+/// Receives incoming [`Connection`]s through [`Stream`].
+#[derive(Clone)]
+pub struct Endpoint {
+    /// Initiate new connections or close socket.
+    endpoint: quinn::Endpoint,
+    /// Receiving new incoming connections.
+    receiver: RecvStream<'static, Result<Connection>>,
+    /// Task handle handling new incoming connections.
+    task: Task<()>,
+}
+
+impl Debug for Endpoint {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        f.debug_struct("Server")
+            .field("endpoint", &self.endpoint)
+            .field("receiver", &String::from("RecvStream<Connection>"))
+            .field("task", &self.task)
+            .finish()
+    }
+}
+
+impl Endpoint {
+    /// Encapsulates common construction paths for
+    /// [`new_server`](Endpoint::new_server) and
+    /// [`new_client`](Endpoint::new_client).
+    ///
+    /// # Errors
+    /// - [`Error::ParseAddress`] if the `address` couldn't be parsed
+    /// - [`Error::MultipleAddresses`] if the `address` contained more then one
+    ///   address
+    /// - [`Error::BindSocket`] if the socket couldn't be bound to the given
+    ///   `address`
+    fn new(
+        address: impl ToSocketAddrs,
+        server_config: ServerConfig,
+        client_config: ClientConfig,
+    ) -> Result<Self> {
+        // parse socket, this can return two different errors, see documentation
+        let address = parse_socket(address)?;
```

Comment:
I originally was coming to make a comment to suggest replacing ToSocketAddrs with tokio's version, but then I realized that `Endpoint::new` isn't async. Since this crate is meant to provide async networking, it feels like we should be assuming that every method being called will be called from an async task. I'd recommend switching endpoint initialization to async.

`ToSocketAddrs` may do DNS resolution, which is why it can be blocking.

Reply:
I didn't realize that you can use DNS resolution inside this function, this wasn't intended.
DNS resolution should happen before you pass a `SocketAddr` in here. Preferably with trust-dns instead of system API. I will put that in the backlog for now.

Reply:
Fixed as discussed.